resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala

```diff
@@ -65,14 +72,31 @@ object ExecutorPodsSnapshot extends Logging {
         case "pending" =>
           PodPending(pod)
         case "running" =>
+          // If we're checking all containers look for any non-zero exits
           if (shouldCheckAllContainers &&
             "Never" == pod.getSpec.getRestartPolicy &&
             pod.getStatus.getContainerStatuses.stream
               .map[ContainerStateTerminated](cs => cs.getState.getTerminated)
               .anyMatch(t => t != null && t.getExitCode != 0)) {
             PodFailed(pod)
           } else {
-            PodRunning(pod)
+            // Otherwise look for the Spark container
+            val sparkContainerStatusOpt = pod.getStatus.getContainerStatuses.asScala
+              .find(_.getName() == sparkContainerName)
+            sparkContainerStatusOpt match {
+              case Some(sparkContainerStatus) =>
+                sparkContainerStatus.getState.getTerminated match {
+                  case null => PodRunning(pod)
+                  case t =>
+                    if (t.getExitCode == 0) {
+                      PodSucceeded(pod)
+                    } else {
+                      PodFailed(pod)
+                    }
+                }
+              // If we can't find the Spark container status, fall back to the pod status
+              case _ => PodRunning(pod)
```

Comment:
When does this happen? This looks like the case, `pod.getStatus.getContainerStatuses.asScala.find(_.getName() == sparkContainerName) == None`. In that case, isn't it a bug?

Reply:
Yeah, I misread the codes. This looks it should be fixed. I thought `spark.kubernetes.executor.podTemplateContainerName` is optional so it's fine but I just realised that we're setting a default value below.

Reply:
So the idea here is that if for some reason we can't find the Spark container defaulting to the pod status is the most reasonable thing to do. The alternative would be to throw an exception here since it indicates something unexpected has happened.

Reply:
In that case, `logWarning` will be more helpful for us to detect it and fix it.

Reply:
Sounds good, I'll add a logWarning to this.